src/AppInstallerRepositoryCore/Microsoft/PreIndexedPackageSourceFactory.cpp

```diff
@@ -389,35 +414,65 @@ namespace AppInstaller::Repository::Microsoft
                 return std::make_shared<DesktopContextSourceReference>(details);
             }
 
-            bool UpdateInternal(const std::string&, Msix::MsixInfo& packageInfo, const SourceDetails& details, IProgressCallback& progress) override
+            bool UpdateInternal(const std::string& packageLocation, Msix::MsixInfo& packageInfo, const SourceDetails& details, IProgressCallback& progress) override
             {
                 // We will extract the manifest and index files directly to this location
                 std::filesystem::path packageState = GetStatePathFromDetails(details);
                 std::filesystem::create_directories(packageState);
 
-                std::filesystem::path manifestPath = packageState / s_PreIndexedPackageSourceFactory_AppxManifestFileName;
-                std::filesystem::path indexPath = packageState / s_PreIndexedPackageSourceFactory_IndexFileName;
+                std::filesystem::path packagePath = packageState / s_PreIndexedPackageSourceFactory_PackageFileName;
 
-                if (std::filesystem::exists(manifestPath) && std::filesystem::exists(indexPath))
+                if (std::filesystem::exists(packagePath))
                 {
-                    // If we already have a manifest, use it to determine if we need to update or not.
-                    if (!packageInfo.IsNewerThan(manifestPath))
+                    // If we already have a trusted index package, use it to determine if we need to update or not.
+                    if (Msix::ValidateMsixTrustInfo(packagePath, WI_IsFlagSet(details.TrustLevel, SourceTrustLevel::StoreOrigin)) &&

```

Comment:
You need another
```
auto indexPackageLock = Utility::ManagedFile::OpenWriteLockedFile(packagePath, 0);
```
to prevent a TOCTOU here as well (between validation and the `IsNewerThan`).

It might even be worth making a helper type like:
```
struct PersistedIndexPackage
{
PersistedIndexPackage(path);
bool ValidateMsixTrustInfo(bool checkMSOrigin);

private:
ManagedFile m_fileLock;
}
```
Then remove the free function `ValidateMsixTrustInfo`.  That way you *must* lock the file to validate it, and hopefully that translates into the appropriate lifetime of the lock in the calling code.